### DISCUSSION TITLE
What are the fan control custom behaviours?

### DISCUSSION THREAD

**Discussion by CharlesMAtkinson** [wiiznokes/fan-control#36] [General]
Hi wiiznokes

Do the fan control custom behaviours include speeding fan(s) according to several temperatures?  My use case is that any one of CPU, GPU and NVMe temperatures may require the most cooling airflow and I would like to speed my single inlet and single outlet fans according to whichever of those currently most needs it.  On Debian Bookworm in case it make any difference

**Comment by wiiznokes** [comment]
Hello,

Currently, all available behaviors of this [picture](https://private-user-images.githubusercontent.com/78230769/237633327-543af76c-137c-456d-a04e-8ebfed323178.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2NDgzMzksIm5iZiI6MTcwMjY0ODAzOSwicGF0aCI6Ii83ODIzMDc2OS8yMzc2MzMzMjctNTQzYWY3NmMtMTM3Yy00NTZkLWEwNGUtOGViZmVkMzIzMTc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE1VDEzNDcxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjODVkMThiODQxNmZmYTgyM2UzNzRmMTFjM2U3OGY2MzY0ZGJhMTkzMzZjYzNjNjg2ZWQ1YmNiZTM0N2IyZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sNscK8neAysrRK0AiXf3YEAdO531dbr_tRSf52DCbZs) are implemented (it's the previous app). I thinks you should be able to do what you want, because you can assign a specific behavior to each controls. You sould try the [prerelease](https://github.com/wiiznokes/fan-control/releases), it have a .deb package

Let me know if it match your use case, or fill an issue about what you miss/want.

Note: a control represente a fan. It is named like that because some fans are not controllable, and it's how computer handle things.

**Comment by CharlesMAtkinson** [comment]
Thanks wiiznokes

The picture link gave a 404.

I installed version 0.1.0 from the .deb.  Here's my notes about what happened after that

Did not find usage instructions in README.md

Is an X11 app which has to be run as root.  Not welcome but I tried it anyway
```
c@cw11:~$ sudo fan-control
[sudo] password for c: 
error: XDG_RUNTIME_DIR is invalid or not set in the environment.
...
```
Window opened.  Xfce's "Window Buttons" showed it with no icon as "Untitled window".  Window had no window decorations so no close etc. buttons.  Found heading "Configuration name" (editable) and columns for PWMs, temperatures and fan speeds.  Clicked around the UI.  Found only possibility to
- Heading: change value
- PWMs: Change name or Delete
- Temperatures: Change name.  Change value to None or one of the other temperatures
- Fan speeds: Change name.  Change value to None or one of the other fan speeds

Changed heading to fan-control-try
Did not find a way to save settings.  Save on exit?
Deleted 0 RPM fans
Closed the window using Xfce's "Window Buttons"
Did not find conffile
```
c@cw11:~$ ll ~/.config/fan-control*
ls: cannot access '/home/c/.config/fan-control*': No such file or directory
c@cw11:~$ find -xdev -iname '*fan-control*'
[no output]
```
Wondered about "it's the previous app" in case that had more usage instructions.  Did not find anything about the previous app at https://github.com/wiiznokes/fan-control including no copyright file in the source code tree

**Comment by wiiznokes** [comment]
> The picture link gave a 404.

Sorry, the image in the read me of https://github.com/wiiznokes/LibreFanControl. This is the previous version of i made.

> Did not find usage instructions in README.md

Ok, thanks for sharing, i will add that latter, there is explanation on each item on the right side of the app tho. But maybe you haven't found it because icons don't works for you (i working on a fix for that) 

> Is an X11 app which has to be run as root. Not welcome but I tried it anyway

I don't see what you meant by that. I thinks the app should works on wayland and x11. It's made with Iced: https://github.com/iced-rs/iced which use winit, and winit support wayland. Iced is still experimental tho, so there is bugs.

>  showed it with no icon as "Untitled window"

I have the same bug. I can't find a fix for that unfortunalltly

> Window had no window decorations so no close etc. buttons

I know, this is the issue: https://github.com/pop-os/libcosmic/issues/225. But we can still press the empty buttons.

>  Found only possibility to

    Heading: change value
    PWMs: Change name or Delete
    Temperatures: Change name. Change value to None or one of the other temperatures
    Fan speeds: Change name. Change value to None or one of the other fan speeds

See the right column, you can add items.

> Did not find a way to save settings

On the top, there are icons but apparently, they don't show up. You can still press them like the close app button, even if there is no ui for them.

> Did not find conffile

Maybe in /root/.config/fan-control ?

**Reply by CharlesMAtkinson** [reply]
Regards the dangers of running X apps (applies equally to Wayland apps) as root https://bugzilla.gnome.org//show_bug.cgi?id=772875#c5

> But maybe you haven't found it because icons don't works for you (i working on a fix for that)

Yes.  Screenshot:
![Screenshot_2023-12-16_09-33-45](https://github.com/wiiznokes/fan-control/assets/22408162/eaaa9d30-59c5-495d-8149-923f08bc1e74)
When fan-control opened, the pointer was captured in adjust width mode.  Only after clicking the left mouse button (I am right handed) did the pointer mode change to selecting items.  Only then was Ctrl+c from the initiating terminal effective to close the window.

Sliding around the top bar found four items, one of which showed a settings cog.  Sliding around the right bar found an add item at the bottom which had a + symbol and a pop-up tool tip
 
> Maybe in /root/.config/fan-control ?

Yes
```
# ll .config/fan-control/
total 8.0K
-rw-r--r-- 1 root root 2.1K Dec 15 18:21 hardware.toml
-rw-r--r-- 1 root root   37 Dec 15 17:27 settings.toml
```
hardware.toml did not show any sign of my having deleted the 0 RPM fans, confirmed by them being visible in the fan-control window

**Reply by wiiznokes** [reply]
> Regards the dangers of running X apps (applies equally to Wayland apps) as root https://bugzilla.gnome.org//show_bug.cgi?id=772875#c5

I partially agree with that. Obviously, it's always better to make an app that doesn't need to be admin to run, but in the case of fan control, on my PC at least, it is impossible to modify the value of a fan without being root. Which makes sense because the files in `/sys/class/hwmon/hwmon6/` are write-protected.

So according to the link provided, I am supposed to create a service which is executed as root and which connects with the app. This is not at all perfect and brings its share of defaults (security, complexity, packaging), the previous app worked this way.

A "perfect" solution in my opinion would be a package similar to flatpak, with permission that allows you to write to /sys/class/hwmon/hwmon6/ (to change pwm values). But that's not about to happen, the domain is moving quite slowly.

For security, this shouldn't pose too many problems I think, especially because the app is coded in Rust, a so-called "safe" language.

> Yes. Screenshot:

You have the icons 😄. Just not the close, maximise and minimise but that normal. If you really want then, you could install https://github.com/pop-os/cosmic-icons but it's better to wait imo (icons will be embed in the executable soon).

To create a new config, write a name, and click on the + button on the top.

> Sliding around the top bar found four items, one of which showed a settings cog. Sliding around the right bar found an add item at the bottom which had a + symbol and a pop-up tool tip

So it works? If there are UI related bug, It's probably because of Iced, not me, it will be fixed with time. But the app should be usable.

> hardware.toml did not show any sign of my having deleted the 0 RPM fans

hardware.toml is regenerated at every launch.
> To create a new config, write a name, and click on the + button on the top.

**Comment by wiiznokes** [comment]
Add added a screenshot on the dev branch: https://github.com/wiiznokes/fan-control/tree/dev.
This is what i get on Windows
